---
title: "DESeq2 Tutorial"
author: "Charlie George & modified by Devika Agarwal"
date: "`r Sys.Date()`"
output: html_document
---

Useful resources 

- **RNAseq Workflow Vignette** - http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
- **Deseq2 Vignette** - http://www.bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

# install packages 

- If i didn't already have these installed in my environment - I would delete the # on lines 19-23 and 26-39 and run the code 

```{r}
# from cran 
# install.packages('tidyverse')
# install.packages("RColorBrewer")
# install.packages('pheatmap')
# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# from bioconductor
# BiocManager::install("ashr")
# BiocManager::install("apeglm")
# BiocManager::install("DESeq2")
# BiocManager::install("org.Mm.eg.db")
```

# load packages

- what do these packages do? Add comments 
```{r}
library("pheatmap")   
library("RColorBrewer")
library(tidyverse)     
library(DESeq2)      
library(ashr)          
library(apeglm)      
library(org.Mm.eg.db)
```


# 1 Prepping our input data to make deseq2dataset object

## Import counts table

- We made this (or a very similar version) using featurecounts in week1 on the cluster
- raw data -> fastqc -> hisat2 -> featurecounts 

```{r}
# Import counts data to variable called featurecounts_table
# The file has a comment line we need to ignore at top of file 
featurecounts_table <- read_tsv("data/allsamples_counts.tsv.gz", comment = "#")


# lets save the gene info (Chr,Start,End,Strand, Length) columns as a seperate dataframe (df) incase we need it later
# call this table gene_info and set row names as Geneid
gene_info <- featurecounts_table %>%
                column_to_rownames("Geneid") %>%                     
                dplyr::select(Chr, Start, End, Strand, Length)


# create a counts table without the gene info (Chr,Start,End,Strand, Length) columns
# to get a clean counts table where:
# rows = genename, columns = samples
counts_table <- featurecounts_table %>%
                  column_to_rownames("Geneid") %>%
                  dplyr::select(-Chr, -Start, -End, -Strand, -Length) %>%
                  as.matrix()



# check the dimensions of this counts_table table using dim() 
# if used pt_Coding only gtf -> -> should have 22,001 genes and 12 samples 
# if use allsamples_counts.tsv.gz they generated on course -> 39,179 genes and 8 cols 
dim(counts_table)
head(counts_table)
```

## Right now lets import and set up our table of sample information

```{r}
sample_info <- read_tsv("data/PRJEB18572_sample_info.tsv")
head(sample_info)

# What class is the cell_type currently? 
class(sample_info$cell_type)



# We want to change it to a factor with 2 levels (the hypothesis we will test)
# Which should be our reference level? 

sample_info$cell_type <- factor(sample_info$cell_type, levels=c("CD4","CD8"))    
#we want the cell_type column to have two levels, but not ordered, order in this function means that the levels are not equivalent (implies a numerical hierarchy which is not present)
# as.factor() will just make it a factor - can't specify the levels (default =alphabetical)


# what other columns might we want to include in our design that we want to change to factors? 
sample_info$genotype <- as.factor(sample_info$genotype) # as.factor makes factor alphabetically 
sample_info$replicate <- as.factor(sample_info$replicate) # as.factor makes factor alphabetically

```

## Check the sample names are the same and in the same order in the sample_info and counts_table 

```{r}
# Column names of the count table and row names of the sample info table must be in the same order
# Check this 

# they don't match 
  # 8 samples in counts_table, 12 in sample_info
  # counts_table has .bam at end 
  
colnames(counts_table) == rownames(sample_info)
colnames(counts_table)

colnames(sample_info)

# make sample_info row names from sample_id 
sample_info <- sample_info %>% 
               column_to_rownames("sample_id") %>% 
               as.data.frame()

# remove .bam from colnames 
colnames(counts_table) <- counts_table %>%
    colnames() %>%
    str_replace(".bam","")

# another solution
#colnames(counts_table) <- gsub(".bam","",colnames(counts_table))



```

# 2. Generate DESeqDataSet object

```{r}
# look up function to use from slides, find help and create an object called dds
dds <- DESeqDataSetFromMatrix(countData = counts_table, colData = sample_info, design = ~ genotype + replicate + cell_type) 
# can put whatever want in design here and adjust later.... 
  
```

## Investigate the dds object 

```{r}
# what does it look like? - e.g. use str(dds)
str(dds)

# how to access counts
counts(dds)

# access colData i.e the sample info table
colData(dds)

# access rowData
rowData(dds)

# access the design
design(dds)

# does it have any rowRanges()?
rowRanges(dds)


# what are its rownames and colnames?
rownames(dds)
colnames(dds)

# what attributes does it have?
attributes(dds)

```

## How to subset a DESeqDataSet Object 

```{r}
# practice subsetting dds object
# select first 3 columns
dds_subset <- dds[,1:3]

```

# Filter out unexpressed genes 

- It's good practice to filter out genes with 0 counts across all samples at the start 
- This reduces the size of the object and makes computations quicker without affecting the results as these genes are filtered out anyway in the analysis 

```{r}
# filter out genes that have 0 counts across all samples and save as an object called `dds_filtered`
dds_filtered <- dds[rowSums(counts(dds))>0, ]
  
# can also be more stringent in filtering depending on the dataset, filter out low expressing genes with less than 10 counts across 3 samples or higher for example
# at least 3 samples with a count of 10 or higher
#keep <- rowSums(counts(dds) >= 10) >= 3
#dds <- dds[keep,]
#nrow(dds)  
  
```

# 3. Perform exploratory analysis on dds_filtered object 

1. transform dds_filtered object using `rlog()` functions
    - name your new variables `rlog_transformed_dds` 
    - we want to set `blind = TRUE` as because we are doing a QC step we do not want the function to take into account the experimental design, we want it to be an unbiased look at how the samples cluster
    - check the class of the rlog_transformed_dds objects 
    - you can look at the tranformed counts table using `assay()` 

NOTE we only use this tranformed data for QC not for downstream analysis or differential expression itself!!


```{r}
#transform data using `rlog()` functions
rlog_transformed_dds <- rlog(dds_filtered, blind=TRUE)


# Notes on `blind` parameter: 
  # blind = TRUE is default
  # blind = TRUE is not the appropriate choice if you expect many or majority of genes to have large differences in counts explainable by   experimental design
  # Mike Love recommends blind = FALSE for visualisation, clustering and machine learning but blind = TRUE for QC
  # Experimental design is only used for calculating the gene-wise dispersion estimates in order to fit a trend line through the dispersions over the mean

  
#check the class of these variable objects 
class(rlog_transformed_dds)
  
#get and look at the tranformed counts table using `assay()`
assay(rlog_transformed_dds) %>% head()

```

## Optional: If you have time do the tranformation again but using the vst() function 

- As a guide 
    - rlog - more robust than VST when size factors vary widely (10x)
    - VST - less sensitive to high count outliers than rlog, quicker - recommended most often 

```{r}
# get vst transformed data using vst() 
vst_transformed_data <-  vst(dds_filtered, blind=TRUE)
```


## Now plot heatmap of sample-sample correlations using rlog transformed count data and cor()


```{r}
# get correlation matrix of the rlog_transformed counts matrix to get correlations between samples 
corr_dds <- cor(assay(rlog_transformed_dds), method="spearman")
corr_dds_pearson <- cor(assay(rlog_transformed_dds), method="pearson")

# use pheatmap to plot the correlation matrix 
pheatmap(corr_dds)
pheatmap(corr_dds_pearson)



```

## try ploting a heatmap of sample distances 

```{r}
# get sample distances -> Annotated what this code is doing?
sampleDists <- dist(t(assay(rlog_transformed_dds))) 
sample_distances_matrix <- as.matrix(sampleDists) 

pretty_blue_colors <- colorRampPalette(rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = pretty_blue_colors)
```

What do you notice here!? what does it tell us about our data? 

Answer: 083/089 and 082/088 are duplicates of one another (their correlation is 1); one in each pair can be removed


## You can subset a dds object and interact with it like you would a DataFrame

```{r}
# remove the two suspect columns using [,] notation - remove suspicious samples 
rlog_transformed_dds_subset <- rlog_transformed_dds[,1:6]

# lets replot the correlation heatmap with the subsetted data - copy code and adjust from above
sampleDists <- dist(t(assay(rlog_transformed_dds_subset))) 
sample_distances_matrix <- as.matrix(sampleDists) 

colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

pheatmap(sample_distances_matrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists, 
         color = colors)

```
What do you notice now about the data and the clustering?

Answer: Samples are now clustering by celltype. 82/83/84 are CD4 and 85/86/87 are CD8


## Now lets use the plotPCA function from DESeq2 to plot PCA of rlog_transformed_data_subset

```{r, fig.height=4, fig.width=6,fig.asp=1}
# plot pca of rlog_transformed_dds using plotPCA()
plotPCA(rlog_transformed_dds_subset, intgroup='cell_type') 



# what does the intgroup parameter do? have a go at changing it...
plotPCA(rlog_transformed_dds_subset, intgroup='cell_type')
plotPCA(rlog_transformed_dds_subset, intgroup='genotype')
plotPCA(rlog_transformed_dds_subset, intgroup='replicate')

# how many variable genes is it using for pca? 
# 500

# plot the pca for all genes 
plotPCA(rlog_transformed_dds_subset, intgroup='sample_name',ntop = nrow(rlog_transformed_dds_subset))
# Selects top genes based on their variance and then uses prcomp with default settings (centering but not scaling) for PCA

```


What does it tell us about our sample?

Answer: 


------------- This is the end of our Exploratory data analysis -----------------

We now go back to our original dds object with the raw counts and can update it with anything we need to from the exploratory analysis 
    e.g. - update design with variables in sample info table we want to regress out 
         - remove samples (poor quaility or those with name switches etc)
         
# 4. Revisiting our inital data for actual differential gene analysis and updating dds object

What do we need to do to our data??? 

## Update dds_filtered object if needed - call it dds_filtered_updated

```{r}

# do we need to remove/edit samples? 
dds_filtered_updated <- dds_filtered[,1:6]
# do we want to update the design? Is this nessecary in this case?
# e.g. design(dds) <- ~replicate + condition
#design(dds_filtered_updated) <- ~cell_type
design(dds_filtered_updated) <- ~replicate + cell_type


```


## Now we can start our actual DE!!!! yay 

## We will run the analysis step by step to see whats going on 

## 1. Estimate size factors using estimateSizeFactors()

```{r}
# Calculate the size factors for each sample using estimateSizeFactors()
dds_filtered_updated <- estimateSizeFactors(dds_filtered_updated)
# Access the size factors from the dds_filtered_updated object
sizeFactors(dds_filtered_updated)
# Do you see any systematic differences in size factors between groups?

```
What does a size factor below one mean? Gene expression values for those samples will be relatively increased - either they have relatively low average library size, or composition in which a few genes are highly expressed so the rest of the genes have been undersampled

## Optional: Plot size factors as barchart using ggplot

```{r}
size_factors <- sizeFactors(dds_filtered_updated)
size_factors <- data.frame(sample = names(size_factors),
                           size_factor = size_factors,
                           sample_group = colData(dds_filtered_updated)$cell_type)

ggplot(size_factors, aes(x = sample, y = size_factor, fill = sample_group)) +
    geom_col() +
    theme(axis.text.x = element_text(angle = 45, h = 1))
```

## Get counts table normalised using size factors

```{r}
# get counts table normalised using size factors - call it normalized_counts
normalized_counts <- counts(dds_filtered_updated, normalized=TRUE)


# check the class of the output
class(normalized_counts)
```

# 3. Estimate dispersions for each gene

```{r}
# Obtain dispersion estimates for each gene using estimateDispersions() 
dds_filtered_updated <- estimateDispersions(dds_filtered_updated)
# Plot the per-gene dispersion estimates (DESeq2 has a helper function for this)
plotDispEsts(dds_filtered_updated )
```

Does it look ok? - extra code to investigate further

```{r}
# Access dispersions from the DESeqDataSet object for individual genes
head(dispersions(dds_filtered_updated))
rowData(dds_filtered_updated)    # mcols(dds) accesses the same slot
View(as.data.frame(rowData(dds_filtered_updated)))

# Description of columns in mcols
mcols(mcols(dds_filtered_updated), use.names = TRUE)


# dimensions of dispersion = one dispersion per gene
# NA values where 0 counts for all samples for that gene
```

## 4. Perform Wald test

```{r}
# Tests for significance of coefficients in a negative binomial GLM, using previously calculated size factors and dispersion estimates and WaldTest
dds_filtered_updated <- nbinomWaldTest(dds_filtered_updated)
results(dds_filtered_updated)
summary(results(dds_filtered_updated))

##### EXTRA ####### 
# if have more than one group in contrast can pull out those comparisions by specifying contrasts 
# check contrast names 
resultsNames(dds_filtered_updated) # to see what names to use
contrast <- list(resultsNames(dds_filtered_updated)[1], resultsNames(dds_filtered_updated)[2])
contrast
```
Contrasts can be supplied as a character vector with exactly three elements: 
  - the name of the factor (of interest) in the design formula, 
  - the name of the two factors levels to compare. 
  - The factor level given last is the base level for the comparison.
  
	contrast <- c("condition", "level_to_compare", "base_level")
	results(dds, contrast = contrast)

# So we have the actual steps that we've gone through to get to DE in one place generate the dds object again and use the DESeq() function to perform steps 1-3 in one go

```{r}

# filter rows with 0 counts 
dds_actual_analysis <- dds[rowSums(counts(dds)) > 0, ]

# remove unwanted samples 
dds_actual_analysis <- dds_actual_analysis[,1:6]

# make DESeqDataSet (dds) object 
design(dds_actual_analysis) <- ~ replicate + cell_type

# run DESeq() 
dds_actual_analysis <- DESeq(dds_actual_analysis)
```

# 7. Get the results from your dds object - call it `results_obj`

```{r}
# What do the results() and summary() do, what arguments can you specify? 
results_obj <- results(dds_actual_analysis, alpha=0.05)
summary(results_obj)
```

## Optional - if we have time: Access the coefficients of the Negative Binomnial GLM – see vignette("DESeq2")
```{r}

head(coef(dds_actual_analysis))
coef_dds <- coef(dds_actual_analysis)

# Intercept is absolute gene expression in baseline group in rea
# Coefficients are the logFC in the other conditions
# Coefficients are the log2 fold changes for each group

##### EXTRA #############################

# Convert coefficients to "real" space rather than log values
coef_real <- 2^(coef(dds_actual_analysis))
coef_real["ENSMUSG00000041779", ]
# Intercept     376.604199
# cell_type_CD8_vs_CD4    1.251456

# Intercept - log2 normalised expression level of gene for baseline condition, which is Egr2_3_DKO_CD4 (uses size factors and dispersion estimates calculated earlier)
```

# Plot normalised counts for gene ENSMUSG00000041779

```{r}

gene_counts_filt <- counts(dds_actual_analysis, normalized = TRUE)
gene_counts_filt <- gene_counts_filt["ENSMUSG00000041779", ]
gene_counts_filt <- data.frame(counts = gene_counts_filt,
                               condition = dds_actual_analysis$cell_type)
gene_counts_filt
ggplot(gene_counts_filt, aes(x = condition, y = counts)) +
    geom_boxplot() +
    geom_point()
# Coefficients above show that gene is expressed at 1.25 times the background level in Egr2/3 DKO CD8s (control is Egr2/3 DKO CD4)
```


```{r}
# Another way to do the same plot using plotCounts functon to returnData 
plot_data <- plotCounts(dds = dds_actual_analysis, gene = "ENSMUSG00000041779", intgroup = c("cell_type","replicate"), returnData = T)
ggplot(plot_data, aes(x = cell_type, y = count, group=cell_type)) +
   geom_boxplot() +
  geom_point(aes(color=replicate), size=2)
```

## Access the results table for the comparison between CD8 and CD4 T cells

```{r}
resultsNames(dds_actual_analysis)
results_CD4vsCD8 <- results(dds_actual_analysis,contrast = c('cell_type','CD4','CD8'))
summary(results_CD4vsCD8)


results_CD8vsCD4 <- results(dds_actual_analysis,contrast = c('cell_type','CD8','CD4'))
summary(results_CD8vsCD4)


results_rep1vsrep3 <- results(dds_actual_analysis,contrast = c('replicate','1','3'))
summary(results_rep1vsrep3)
```

Table should contain log2 fold changes and p-values for all genes

## Plot a histogram of the raw and BH-adjusted p-values using ggplot – do they look as expected?

```{r}
results_obj %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_histogram(aes(x=pvalue), fill='lightblue') + 
    geom_histogram(aes(x=padj),fill='lightgreen')

```


## Generate an MA plot of the log2 FC values for all genes using DESeq2 plotMA()

```{r}
plotMA(results_obj,alpha=0.05, colNonSig = "gray60")

results_obj %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 
results_obj


```

## Shrink the log2 FC values using the normal, apeglm and ashr methods -> see the vignette – which shall we choose? 

```{r}

# call your results table  called shrunken_dds_results_df
shrunken_dds_results_apeglm <- lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="apeglm")

#using the DEseq MA plot function
plotMA(shrunken_dds_results_apeglm, alpha=0.05, colNonSig = "gray60",
  colSig = "blue",
  colLine = "black",)

#plotting the same graph in ggplot
shrunken_dds_results_apeglm %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

```


```{r}
# test otehr shrinkage methods
#plotting the same graph in ggplot
lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="normal") %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

# test otehr shrinkage methods
# plotting the same graph in ggplot
lfcShrink(dds_actual_analysis, coef = "cell_type_CD8_vs_CD4", type="ashr") %>% 
    as.data.frame() %>% 
    ggplot() +
    geom_point(aes(x=log2(baseMean),y=log2FoldChange,color=padj < 0.05),size=0.5) 

```

## Generate a results table (one shrinkage method) containing mgi symbols

```{r}
# Use the org.Mm.eg.db package (hint:you did this yesterday)
# Remove all genes with a padj of NA
# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?

keytypes(org.Mm.eg.db)      #to remind us of what keytypes are in this particular db

# # mapIds gives you a 1:1 mapping by default (can only extract one column at a time), whereas the select function doesn't (but you can extract multiple columns at a time)

MGI_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "MGI", "ENSEMBL")
gene_symbols <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "SYMBOL", "ENSEMBL")
entrez_ids <- mapIds(org.Mm.eg.db, row.names(shrunken_dds_results_apeglm), "ENTREZID", "ENSEMBL")

table(names(entrez_ids)==row.names(shrunken_dds_results_apeglm))     #to check that the order of the vectors are the same as the df
#options(scipen=0)

#add mapId vectors to the original results df
results_shrunken_LFC_apeglm_filtered_df <- shrunken_dds_results_apeglm %>%
    as.data.frame() %>% 
    mutate(mgi_symbols=MGI_symbols) %>%
    mutate(gene_symbols=gene_symbols) %>%
    mutate(ENTREZ_IDs=entrez_ids)


# How many Ensembl IDs are not assigned an mgi symbol and how many mgi symbols are duplicated?
which(duplicated(names(MGI_symbols)))

```


## Export the results to CSV

```{r}
results_shrunken_LFC_apeglm_filtered_df <- results_shrunken_LFC_apeglm_filtered_df %>% 
						rownames_to_column("ensembl_gene_ids")
						
write_csv(results_shrunken_LFC_apeglm_filtered_df, "CD8_vs_CD4_DE_results_lfcshrinkage.csv")

```




Well done! That is the basics of exploratory data analysis for bulk RNAseq and differential gene expression - these statistical methods underly the foundations for single cell differential expression analysis, although scRNA-seq has a few extra paramterisations and complications to take into account. 

We will now progress to pathway/geneset analysis to get some biological meaning from our DE gene list....


Bonus Exercise 

Generate a heatmap of normalised counts for the top 20 (by shrunken FC) differentially-expressed genes across all samples– label samples by condition and genes by mgi symbol
```{r}
# get the results table for 20 genes with the highest log2 fold changes from shrunken lfc results object  
top_20_DE_genes_df <- results_shrunken_LFC_apeglm_filtered_df %>% 
                                dplyr::filter(padj < 0.05) %>% 
                                arrange(desc(abs(log2FoldChange))) %>% 
                                head(20)

top_20_DE_genes_df

# get the ensembl ids of these genes 
top_20_DE_gene_ids = top_20_DE_genes_df$ensembl_gene_ids
# get the gene symbols for these genes 
top_20_DE_gene_symbols = top_20_DE_genes_df$gene_symbols

# get normalised counts matrix for the top 20 DE genes 
norm_counts_for_top20_de_genes <- normalized_counts[top_20_DE_gene_ids,]
# set mgi symbols as row names 
row.names(norm_counts_for_top20_de_genes) <- top_20_DE_gene_symbols

# make heatmap of normalised counts
pheatmap(norm_counts_for_top20_de_genes, 
         cluster_columns = FALSE, 
         scale = "row")



```


Plot a volcano plot:
Highlight significantly differentially expressed genes (p adj < 0.05, log2FC > 2) in red
Add labels to highlight the location of some interesting genes

```{r}
# volcano plot has fold change on x-axis and -log10(padj) on y axis 
# create a new column for -log10(padj)
# create a new column to label genes with TRUE/FALSE if log2foldchange > 2 & padj < 0.05 
df_for_volcano_plot <- results_shrunken_LFC_apeglm_filtered_df %>% 
                                                    filter(!is.na(padj)) %>%           # remove genes not tested in DE (padj == NA) %>% 
                                                    mutate(log10p = -log10(padj)) %>%  # make a column for -log10(padj)
                                                    mutate('sig_gene' = if_else(padj<0.05 & abs(log2FoldChange) > 2,TRUE,FALSE)) # add true and false to using if_else(condition_to_test,value_if_true,value_if_false)



# volcano plot has fold change on x-axis and -log10(padj) on y axis 
volcano_plot <- df_for_volcano_plot %>% 
                            ggplot() +                                                        # plot
                            geom_point(aes(x = log2FoldChange, y = log10p, color=sig_gene)) + # add points colored by sig_gene col (Lfc >2, padj<0.05)
                            scale_color_manual(values = c('TRUE' = "red", 'FALSE' = "black")) # set colors manually 

volcano_plot
# Note the y-axis is 'capped' at ~ 300 - these are genes with very very small padj values - hard to display thier scale with the rest of the data 

# Can add labels of interesting genes using geom_label e.g. 


  
 volcano_plot 
  
  
 # If we wanted to subset/filter data for Cd4 gene we could do subset(df_for_volcano_plot, gene_symbols == 'Cd4')
 # we could give subset a list of genes of interest that we know about 
 volcano_plot + 
    geom_label(data=subset(df_for_volcano_plot, gene_symbols == c('Cd4','Cd8a','Actb')),
               aes(x = log2FoldChange, y = log10p, label=gene_symbols))
 
 
 
# or we could label the sig genes with the top 20 highest foldchanges 
 top_20_highest_foldchange <-  df_for_volcano_plot %>% 
                                dplyr::filter(padj < 0.05) %>% 
                                arrange(desc(abs(log2FoldChange))) %>% 
                                head(20)
 
 volcano_plot + 
    geom_label(data=top_20_highest_foldchange,
               aes(x = log2FoldChange, y = log10p, label=gene_symbols))

 # in this case the labels are harder to see - ggrepel library has a nicer labelling fuction to avoid overlapping labels but you can see it hides some that overlap too much - you can play around with the settings to get more to show
  volcano_plot + 
    ggrepel::geom_text_repel(data=top_20_highest_foldchange,
               aes(x = log2FoldChange, y = log10p, label=gene_symbols))

 
```
Output the sessionInfo() so know what packages we used
```{r}
sessionInfo()

```


---------------------------   This extra code might come in handy if want to extend the tutorial -----------


## Filter res_df and save as CSV

```{r}
#Filter the results table for padj < 0.05 and absolute log2 FC > 1, and write to a CSV file
res_0.05 <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & abs(log2FoldChange) > 1)

#Get upregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC > 1, and write to a CSV file
res_0.05_up <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange > 1)

#Get doownregulated genes - Filter the results table for padj < 0.05 and log2FoldChange FC < 1, and write to a CSV file
res_0.05_down <- results_shrunken_LFC_apeglm_filtered_df %>% 
    dplyr::filter(padj < 0.05 & log2FoldChange < -1)

write_csv(res_0.05, 
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1.csv")

write_csv(res_0.05_up,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_up.csv")

write_csv(res_0.05_down,
          "CD8_vs_CD4_DKO_DE_results_lfcshrinkage_0.05_log2FC1_down.csv")
```


# Volcano plot

```{r}
# volcano_df <- as.data.frame(res_apeglm) %>% 
#     rownames_to_column(var = "Ensembl_ID") %>% 
#     left_join(geneIDs, by = c("Ensembl_ID" = "GENEID")) %>%
#     dplyr::filter(!is.na(padj)) %>% 
#     mutate(log10p = -log10(padj))

volcano_df <- res_df %>% 
    mutate(log10p = -log10(padj)) %>% 
    arrange(desc(log10p))

my_colours <- c("red", "black")
names(my_colours) <- c("TRUE", "FALSE")
my_colours <- my_colours[as.character(ifelse(volcano_df$padj < 0.05 & abs(volcano_df$log2FoldChange) > 1, TRUE, FALSE))]

ggplot(volcano_df, aes(x = log2FoldChange, y = log10p)) +
    geom_point(colour = my_colours) +
    geom_label(data = volcano_df[which(volcano_df$mgi_symbol == "Cd4"), ], aes(label = "Cd4")) +
    geom_label(data = volcano_df[which(volcano_df$mgi_symbol == "Cd8a"), ], aes(label = "Cd8a"))

ggplot(volcano_df, aes(x = log2FoldChange, y = log10p)) +
    geom_point(colour = my_colours) +
    geom_label_repel(data = volcano_df %>% slice_head(n = 20), aes(label = mgi_symbol[1:20]))

ggplot(volcano_df, aes(x = log2FoldChange, y = log10p)) +
    geom_point(colour = my_colours) +
    geom_text_repel(data = volcano_df %>% slice_head(n = 20), aes(label = mgi_symbol[1:20]))
```

# Other plots

```{r}
# plotCounts() function normalises counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting
plotCounts(dds, gene = which.min(res$padj), intgroup = "condition")     # Ikzf2
plotCounts(dds, gene = "ENSMUSG00000023274", intgroup = "condition")     # Cd4
plotCounts(dds, gene = "ENSMUSG00000053977", intgroup = "condition")     # Cd8a
```

```{r}
# Plotting with ggplot
d <- plotCounts(dds, gene = "ENSMUSG00000023274", intgroup = "condition", 
                returnData = TRUE)

ggplot(d, aes(x = condition, y = count, colour = condition)) + 
    geom_point(position = position_jitter(w = 0.1, h = 0)) + 
    scale_y_log10(breaks = c(25, 100, 400)) +
    theme_cowplot() +
    theme(axis.text.x = element_text(angle = 45, h = 1))
```

# Different model - test for differences between DKO and Kin mice controlling for differences in cell type

```{r}
sample_info$condition <- gsub("Egr2_3", "Egr2.3", sample_info$condition)
sample_info2 <- sample_info %>% 
    separate(condition, into = c("gene", "genotype", "celltype"), sep = "_") %>% 
    unite("genotype", gene, genotype)

dds_genotype <- DESeqDataSetFromMatrix(
    countData = count_table,
    colData = sample_info2,
    design = ~ celltype + genotype)
design(dds_genotype)
dds_genotype <- DESeq(dds_genotype)

# Model coefficients
colnames(coef(dds_genotype))
head(coef(dds_genotype))
head(2 ^ (coef(dds_genotype)))

# Results
resultsNames(dds_genotype)
results_KO_Kin <- results(dds_genotype, name = "genotype_Egr2.3_DKO_vs_Egr2_Kin")
```

# Different model - test for interaction effects between genotype and cell type

```{r}
dds_interaction <- DESeqDataSetFromMatrix(
    countData = count_table,
    colData = sample_info2,
    design = ~ celltype + genotype + celltype:genotype)
design(dds_interaction)
dds_interaction <- DESeq(dds_interaction)

levels(colData(dds_interaction)$celltype)
levels(colData(dds_interaction)$genotype)

dds_interaction$genotype <- relevel(dds_interaction$genotype, "Egr2.3_DKO")

# Model coefficients
head(coef(dds_interaction))
head(2 ^ (coef(dds_interaction)))
# ENSMUSG00000033845
# Intercept = 2558.4074538 (expression in reference genotype and cell type - CD4 Kin)
# genotype_Egr2.3_DKO_vs_Egr2_Kin = 0.5286905
# celltype_CD8_vs_CD4 = 1.1053986
# celltypeCD8.genotypeEgr2.3_DKO (contains numerator factor levels) = 1.0184265

# celltypeCD8.genotypeEgr2.3_DKO = interaction term - additional effect genotype has on the comparison between CD8 and CD4 cells

# Results
resultsNames(dds_interaction)
results_interaction <- results(dds_interaction,
                               name = "celltypeCD8.genotypeEgr2.3_DKO")
results_interaction <- as.data.frame(results_interaction)
length(which(results_interaction$padj < 0.05 & abs(results_interaction$log2FoldChange) > 1))
# 1691

# ~genotype + condition + genotype:condition

# # the condition effect for genotype I (the main effect)
# results(dds, contrast=c("condition","B","A"))
# 
# # the condition effect for genotype III.
# # this is the main effect *plus* the interaction term
# # (the extra condition effect in genotype III compared to genotype I).
# results(dds, contrast=list( c("condition_B_vs_A","genotypeIII.conditionB") ))
#  
# # the interaction term for condition effect in genotype III vs genotype I.
# # this tests if the condition effect is different in III compared to I
# results(dds, name="genotypeIII.conditionB")
# 
# # the interaction term for condition effect in genotype III vs genotype II.
# # this tests if the condition effect is different in III compared to II
# results(dds, contrast=list("genotypeIII.conditionB", "genotypeII.conditionB"))
```


# Heatmap of top 20 differentially-expressed genes

```{r}
# Using results table which has had log2FC shrinkage (apelgm method)
res_apeglm_top20 <- as.data.frame(res_apeglm) %>% 
    rownames_to_column(var = "Ensembl_ID") %>% 
    dplyr::filter(padj < 0.05) %>% 
    top_n(20, wt = log2FoldChange) %>% 
    pull(Ensembl_ID)

heatmap_df <- as.data.frame(assay(vsd)[rownames(assay(vsd)) %in% res_apeglm_top20, ])
colnames(heatmap_df) <- paste(sample_info$condition, sample_info$replicate, sep = "_")

# mgi_symbol <- mapIds(edb,  keys = rownames(heatmap_df), column = "GENENAME", keytype = "GENEID")
# mgi_symbol[is.na(mgi_symbol)] <- "NA"
# rownames(heatmap_df) <- mgi_symbol

# heatmap_df <- heatmap_df %>% 
#     rownames_to_column(var = "Ensembl_ID") %>% 
#     left_join(geneIDs, by = c("Ensembl_ID" = "GENEID"))
# 
# heatmap_df[is.na(heatmap_df)] <- "NA"
# rownames(heatmap_df) <- heatmap_df$SYMBOL
# heatmap_df <- heatmap_df %>% 
#     dplyr::select(-c(Ensembl_ID, SYMBOL))

mgi_symbol_select <- mgi_symbol[mgi_symbol$EnsemblID %in% rownames(heatmap_df), "mgi_symbol"]
mgi_symbol_select[6] <- "Unknown_1"
mgi_symbol_select[11] <- "Unknown_2"
rownames(heatmap_df) <- mgi_symbol_select

column_annotation <- data.frame(sample = c(rep("CD4 DKO", 3), 
                                           rep("CD8 DKO", 3),
                                           rep("CD4 Kin", 3),
                                           rep("CD8 Kin", 3)),
                                row.names = colnames(heatmap_df))

pheatmap(heatmap_df, cluster_rows = FALSE, scale = "row",
         annotation_col = column_annotation)
```

Things I could add:
- Different dataset with more complex model (ALL dataset?)
- How to import results from Kallisto using tximport
- pcaExplorer to visualisation of DESeq2 output - PCA plots, boxplots of counts
