---
title: "Example code for R statistics"
author: "Kevin Rue-Albrecht"
date: "03/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(readxl)
```

# Demo

## Mean and standard deviation

The base R functions `mean()` and `sd()` compute the mean and standard deviation of a distribution

To demonstrate, let us first generate a vector of random, normally distributed, values.

```{r}
set.seed(1)
x <- rnorm(n = 100, mean = 2, sd = 5)
```

We can then use that vector to demonstrate the functions.

```{r}
mean(x)
```

```{r}
sd(x)
```


**What are optional arguments for those functions?**

> Answer:
> `mean(x, trim = 0, na.rm = FALSE, ...)`
> - Argument `trim` indicates the fraction of observations to trim from each
>   end of the vector `x` before computing the mean.
>   Trimming mean truncating those values to the nearest endpoint,
>   _not_ removing those values!
> - Argument `na.rm` indicates whether to remove `NA` values before computing
>   the mean.
>   The presence of any `NA` value in `x` causes the return value to be `NA`.
> `sd(x, na.rm = FALSE)`
> - Argument `na.rm` has the same meaning as above.
>   The same consequences too, in the presence of any `NA` value.

**Why do you think the mean and standard deviation are not exactly those that we would expect?**

> Answer:
> The vector of random values produced by `rnorm` are simulated following the
> normal distribution.
> However, once generated, the values are not refined to ensure an exact match
> to the requested mean and standard deviation.
> As a result, the mean and standard deviation of simulated values is generally
> close but not absolutely equal to the requested values.

# Exercise

## Generate and visualise a distribution

- Generate a vector of 1,000 normally distributed values with mean 10 and standard deviation 5.

```{r}
set.seed(1)
x <- rnorm(n = 1000, mean = 10, sd = 5)
head(x)
```

- Inspect the output of the `summary()` function for that vector.

```{r}
summary(x)
```

- Compute the mean and standard deviation for those values.

```{r}
mean(x)
```

```{r}
sd(x)
```

- Compute the deciles (i.e. 10 evenly spaced quantiles) for those values.

```{r}
quantile(x = x, probs = seq(0, 1, 0.1))
```

- Visualise the distribution of those values as a histogram.

```{r}
hist(x, breaks = 50)
```

- Visualise as vertical lines on the histogram: the mean (red solid), median (red dashed), one standard deviation from the mean (blue solid), and one median absolute deviation from the median (blue dashed).

```{r}
hist(x, breaks = 50)
abline(v = mean(x), col = "red", lty = 1)
abline(v = median(x), col = "blue", lty = 1)
abline(v = mean(x) + sd(x) * c(-1, 1), col = "red", lty = 2)
abline(v = median(x) + mad(x) * c(-1, 1), col = "blue", lty = 2)
```

- Generate a new vector with _a lot_ more values (e.g., one million).
  Draw again a histogram.
  How does the distribution compare with more data points?

```{r}
set.seed(1)
x_million <- rnorm(n = 1E6, mean = 10, sd = 5)
hist(x_million, breaks = 50)
```

## Query distributions and probabilities

For the standard normal distribution ${\mathcal {N}}(\mu=0 ,\sigma ^{2}=1)$:

- Plot the cumulative distribution function in the range $[-5, 5]$ in 0.1 increment.

```{r}
x <- seq(from = -5, to = 5, by = 0.1)
p <- pnorm(q = x, mean = 0, sd = 1)
plot(x = x, y = p)
```

- Plot the inverse cumulative distribution function for quantiles in 0.01 increment.

```{r}
p <- seq(from = 0, to = 1, by = 0.01)
x <- qnorm(p = p, mean = 0, sd = 1)
plot(x = p, y = x)
```

- Plot the density function in the range $[-5, 5]$ in 0.1 increment.

```{r}
x <- seq(from = -5, to = 5, by = 0.1)
d <- dnorm(x = x, mean = 0, sd = 1)
plot(x = x, y = d)
```

- What is the probability of observing a value greater than 2?

```{r}
1 - pnorm(q = 2, mean = 0, sd = 1)
```

- What is the probability of observing a value between -2 and 2?

```{r}
pnorm(q = 2, mean = 0, sd = 1) - pnorm(q = -2, mean = 0, sd = 1)
```

- What is the probability of observing a value more extreme than -2 or 2?

```{r}
1 - (pnorm(q = 2, mean = 0, sd = 1) - pnorm(q = -2, mean = 0, sd = 1))
```

# Demo

## Empirical Cumulative Distribution Function

```{r}
ecdf_iris_sepal_length <- ecdf(iris$Sepal.Length)
ecdf_iris_sepal_length
```

```{r}
ggplot(iris, aes(Sepal.Length)) +
  geom_histogram(color = "black", fill = "grey") +
  cowplot::theme_cowplot()
```

```{r}
plot(ecdf_iris_sepal_length, cex = 0.5)
```

# Demo

## ecdf - Knots

```{r}
knots(ecdf_iris_sepal_length)
```

```{r}
sort(unique(iris$Sepal.Length))
```

# Demo

## ecdf - Quantiles

```{r}
quantile(ecdf_iris_sepal_length, c(0, 0.25, 0.5, 0.75, 1))
```

```{r}
quantile(iris$Sepal.Length, c(0, 0.25, 0.5, 0.75, 1))
```

# Demo

## <i class="fab fa-r-project"></i> Functions for Statistical Testing

```{r}
?pairwise.t.test
```

```{r}
help(pairwise.t.test)
```

# Demo

## Parametric t-test

Two normal distributions.

```{r}
set.seed(1)
x <- rnorm(n = 1000, mean = 0, sd = 1)
y <- rnorm(n = 1000, mean = 1, sd = 1)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
ggplot(test_data, aes(value)) +
  geom_histogram(fill = "grey", color = "black") +
  facet_wrap(~group, ncol = 1) +
  cowplot::theme_cowplot()
```

Unpaired t-test.

```{r}
t.test(value ~ group, test_data)
```

Compare with

```{r}
t.test(x, y)
```

```{r}
t.test(y, x)
```

# Demo

## Non-parametric wilcoxon test

Two uniform distributions

```{r}
set.seed(1)
x <- runif(n = 1000, min = 1, max = 11)
y <- runif(n = 1000, min = 3, max = 13)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```
]

Mann-Whitney U test

```{r}
wilcox.test(value ~ group, test_data)
```

Directed hypothesis

```{r}
wilcox.test(value ~ group, test_data, alternative = "less")
```

# Demo

## Paired test

For each sample, the two measurements are related to one another; e.g. patients measured before and after a treatment.

```{r}
set.seed(1)
n_sample <- 10
x <- runif(n = n_sample, min = 10, max = 20)
y <- x + 2 + rnorm(n = n_sample, mean = 0, sd = 1)
```

```{r}
test_data <- tibble(
  sample = paste("sample", seq_len(n_sample)),
  x = x,
  y = y
) %>% pivot_longer(cols = c(x, y), names_to = "variable")
```

```{r}
ggplot(test_data, aes(variable, value)) +
  geom_line(aes(group = sample), size = 0.1) +
  geom_point() +
  cowplot::theme_cowplot()
```

```{r}
t.test(x, y, paired = TRUE)
```

# Demo

## Analysis of Variance (ANOVA)

```{r}
set.seed(1)
n_sample <- 1000
x1 <- rnorm(n = n_sample, mean = 10, sd = 2)
x2 <- x1 + 5 + rnorm(n = n_sample, mean = 0, sd = 1)
x3 <- x2 + 0 + rnorm(n = n_sample, mean = 0, sd = 0.5)
test_data <- bind_rows(
  tibble(group = "x1", value = x1),
  tibble(group = "x2", value = x2),
  tibble(group = "x3", value = x3)
)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x1", value = x1),
  tibble(group = "x2", value = x2),
  tibble(group = "x3", value = x3)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```

```{r}
out <- aov(value ~ group, test_data)
out
```

```{r}
summary(out)
```

# Demo

## Linear models

```{r}
set.seed(1)
test_data <- tibble(
  x = rnorm(n = 50, mean = 0, sd = 1),
  y = 10 + 2.5 * x + rnorm(n = 50, mean = 0, sd = 0.5))
```

```{r}
ggplot(test_data, aes(x = x, y = y)) +
  geom_point() +
  stat_smooth(method = "glm", se = FALSE) +
  cowplot::theme_cowplot()
```

```{r}
lm(y ~ x, test_data)
```

# Demo

## Linear models - summary

```{r}
lm(y ~ x, test_data) %>% summary()
```

# Demo

## Fisher's Exact Test

```{r}
x_table <- matrix(data = c(12, 4, 3, 23),
  nrow = 2,
  dimnames = list(
    c("DE","Not DE"),
    c("in_pathway", "not_pathway")
  ))
knitr::kable(x_table)
```

```{r}
fisher.test(x_table)
```

# Demo

## Beware of interpreting inadequate tests!

Two uniform distributions.

```{r}
set.seed(1)
n_size <- 10E3
x <- runif(n = n_size, min = 1, max = 11)
y <- runif(n = n_size, min = 3, max = 13)
```

```{r}
test_data <- bind_rows(
  tibble(group = "x", value = x),
  tibble(group = "y", value = y)
)
gg <- ggplot(test_data, aes(value)) +
  facet_wrap(~group, ncol = 1) +
  geom_histogram(fill = "grey", color = "black") +
  cowplot::theme_cowplot()
gg
```

Nothing prevents users from running a parametric test.

```{r}
t.test(value ~ group, test_data)
```

# Demo

## Knowledge assumptions - Central tendency

```{r, message=FALSE, warning=FALSE}
x_mean <- 0
x_sd <- 20
data_table <- tibble(x = as.integer(rnorm(n = 10E3, mean = x_mean, sd = x_sd)))
summary_table <- bind_rows(
  tibble(Value = "mean", value = mean(data_table$x)),
  tibble(Value = "median", value = median(data_table$x)),
  tibble(Value = "mode", value = as.integer(names(which.max(table(data_table$x)))))
)
data_table %>% 
  ggplot() +
  geom_histogram(aes(x = x), color = "black", fill = "grey") +
  geom_vline(aes(xintercept = value, color = Value), summary_table, size = 2, alpha = 0.3) +
  cowplot::theme_cowplot()
```

# Demo

## Knowledge assumptions - Normality

Normal distribution

```{r}
x <- rnorm(n = 5000, mean = 0, sd = 1)
```

```{r}
ggplot(tibble(x=x)) +
  geom_histogram(aes(x), fill = "grey", color = "black", bins = 20) +
  cowplot::theme_cowplot()
```

```{r}
shapiro.test(x)
```

Log-normal distribution

```{r}
x <- 2^rnorm(n = 5000, mean = 0, sd = 1)
```

```{r}
ggplot(tibble(x=x)) +
  geom_histogram(aes(x), fill = "grey", color = "black", bins = 20) +
  cowplot::theme_cowplot()
```

```{r}
shapiro.test(x)
```

# Demo

## Knowledge assumptions - Normality

Normal distribution

```{r}
x <- rnorm(n = 5000, mean = 5, sd = 3)
qqnorm(x)
```

Log-normal distribution

```{r}
x <- 2^rnorm(n = 5000, mean = 0, sd = 1)
qqnorm(x)
```

# Demo

## Multiple-testing correction

For 1,000 iterations, generate two vectors of normally distributed values with mean 0 and standard deviation 1, and run a t-test to compare the two.

```{r}
set.seed(1)
n_tests <- 1000
compute_p_value <- function(dummy) {
  x <- rnorm(n = 100, mean = 0, sd = 1)
  y <- rnorm(n = 100, mean = 0, sd = 1)
  out <- t.test(x, y)
  out$p.value
}
result_table <- tibble(
  pvalue = vapply(X = seq_len(n_tests), FUN = compute_p_value, FUN.VALUE = numeric(1)),
  BH = p.adjust(p = pvalue, method = "BH"),
  bonferroni = p.adjust(p = pvalue, method = "bonferroni")
)
```

Illustrate one iteration.

```{r, include=TRUE, echo=FALSE, fig.height=3}
data_table <- tibble(
  x = rnorm(n = 100, mean = 0, sd = 1),
  y = rnorm(n = 100, mean = 0, sd = 1)
) %>% pivot_longer(cols = c(x, y))
ggplot(data_table) +
  geom_boxplot(aes(name, value)) +
  geom_jitter(aes(name, value), width = 0.1)
```

Plot the distribution of raw p-values.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(pvalue), fill = "grey", color = "black", breaks = seq(0, 1, 0.05)) +
  scale_x_continuous(limits = c(0, 1)) +
  labs(title = "Raw p-value")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i out of %i raw p-values smaller than 0.05", sum(result_table$pvalue < 0.05), n_tests))
```

Plot corrected p-values using BH correction.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(BH), fill = "grey", color = "black", bins = 20) +
  coord_cartesian(xlim = c(0, 1)) +
  labs(title = "BH correction")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i BH-corrected p-values smaller than 0.05", sum(result_table$BH < 0.05)))
```

Plot corrected p-values using Bonferroni correction.

```{r, include=TRUE, echo=FALSE, fig.height=3}
ggplot(result_table) +
  geom_histogram(aes(bonferroni), fill = "grey", color = "black", bins = 20) +
  coord_cartesian(xlim = c(0, 1)) +
  labs(title = "bonferroni correction")
```

Print a short report message.

```{r, include=TRUE, echo=FALSE, results='asis'}
cat(sprintf("There are %i bonferonni corrected p-values smaller than 0.05", sum(result_table$bonferroni < 0.05)))
```

# Exercise

## Statistical tests

### Wilcoxon test

- Run a Wilcoxon test to compare the vectors `1:10` and `5:14`.

```{r}
wilcox.test(x = 1:10, y = 5:14)
```

**What is the p-value?**

> Answer:
> 0.01701

### T-test

- Import the data in the file `gene_exprs.csv`.

```{r}
df <- read.csv("data/gene_exprs.csv")
df
```


- Run a t-test to compare the gene expression values between the two groups.

```{r}
t.test(gene_exprs ~ group, df)
```

**What is the p-value?**

> Answer:
> 0.0182

# Exercise

## Testing & Multiple testing correction

Given an Excel file that contains a matrix of log-normalised counts (`logcounts`) and experimental metadata (`cell_info`),
test each gene (i.e., row) in the matrix for differential expression between the two experimental groups.
Start by importing the `logcounts` table and converting it to a matrix.

```{r}
logcounts_matrix <- read_excel("data/GSE111543.xlsx", "logcounts") %>% 
  column_to_rownames("gene") %>% 
  as.matrix()
dim(logcounts_matrix)
```

```{r}
cell_info <- read_excel("data/GSE111543.xlsx", "cell_info")
# rownames(cell_info) <- cell_info$Sample
# all(rownames(cell_info) == colnames(logcounts_matrix))
head(cell_info)
```

### Approach

1. Write the code to test a single gene and access the p-value.

```{r}
gene_index <- 1
test_data <- data.frame(
    value = as.numeric(logcounts_matrix[gene_index, ]),
    group = cell_info$Infection
)
out <- t.test(value ~ group, test_data)
out$p.value
```

2. Write a function that generalises the code to test any one gene and return the p-value.

```{r}
t_test_row <- function(index, matrix, group) {
    test_data <- data.frame(
        value = as.numeric(matrix[index, ]),
        group = group
    )
    out <- t.test(value ~ group, test_data)
    out$p.value
}
t_test_row(index = 1, matrix = logcounts_matrix, group = cell_info$Infection)
```

3. Use the function `vapply` to test every row in the matrix and collect a vector of p-values.

```{r}
t_test_pvalues <- vapply(
  X = seq_len(nrow(logcounts_matrix)),
  FUN = t_test_row,
  FUN.VALUE = numeric(1),
  matrix = logcounts_matrix,
  group = cell_info$Infection)
names(t_test_pvalues) <- rownames(logcounts_matrix)
head(t_test_pvalues)
```

### Bonus points

- Visualise a histogram of the p-values.

```{r}
hist(t_test_pvalues, breaks = 100)
```

- Correct p-values for multiple testing.
  How many genes remain before and after multiple testing?

```{r}
result_bh <- p.adjust(t_test_pvalues, method = "BH")
table(result_bh < 0.05)
```

- Use `gene_info` to get the gene name for the gene identifier with the smallest p-value.

```{r}
gene_table <- read_excel("data/GSE111543.xlsx", "gene_info")
top_gene_id <- names(result_bh)[which.min(result_bh)]
gene_table$gene_name[gene_table$gene_id == top_gene_id]
```

# Exercise

## Over-representation analysis (ORA)

Given the list of genes (Ensembl gene identifiers) that your identified as differentially expressed in the previous exercise,
and a list of gene sets (`go_db`),
test each gene set for over-representation of differentially expressed genes.
Start by importing the Gene Ontology table and converting it to a list.

```{r}
go_table <- read_excel("data/GSE111543.xlsx", "go_db")
go_list <- split(go_table$ensembl_gene_id, go_table$go_id)
go_list[1]
```

### Approach

1. Write the code to test a single gene set and access the p-value.

```{r}
query <- names(result_bh)[result_bh < 0.05]
query <- na.omit(query)
universe <- rownames(logcounts_matrix)
geneset <- go_list[[1]]
cross_table <- data.frame(
  gene_id = universe,
  geneset = factor(universe %in% geneset, c(TRUE, FALSE)),
  query = factor(universe %in% query, c(TRUE, FALSE))
)
test_table <- table(cross_table$geneset, cross_table$query)
out <- fisher.test(test_table)
out$p.value
```

2. Write a function that generalises the code to test any one gene set and return the p-value.

```{r}
fisher_test_pathway <- function(index, pathways, query, universe) {
  query <- na.omit(query)
  geneset <- pathways[[index]]
  cross_table <- data.frame(
    gene_id = universe,
    geneset = factor(universe %in% geneset, c(TRUE, FALSE)),
    query = factor(universe %in% query, c(TRUE, FALSE))
  )
  test_table <- table(cross_table$geneset, cross_table$query)
  out <- fisher.test(test_table)
  out$p.value
}
fisher_test_pathway(index = 1,
  pathways = go_list,
  query = names(result_bh)[result_bh < 0.05],
  universe = rownames(logcounts_matrix))
```

3. Use the function `vapply` to test every gene set in the list and collect a vector of p-values.

```{r}
fisher_test_pvalues <- vapply(
  X = seq_len(length(go_list)),
  FUN = fisher_test_pathway,
  FUN.VALUE = numeric(1),
  pathways = go_list,
  query = names(result_bh)[result_bh < 0.05],
  universe = rownames(logcounts_matrix))
names(fisher_test_pvalues) <- names(go_list)
head(fisher_test_pvalues)
```

### Bonus points

- Visualise a bar plot of the p-values.

```{r}
hist(fisher_test_pvalues, breaks = 100)
```

- Correct p-values for multiple testing.
  How many gene sets remain before and after multiple testing?

```{r}
fisher_test_bh <- p.adjust(fisher_test_pvalues, method = "BH")
table(fisher_test_bh < 0.05)
```

- Use `go_info` to annotate each GO gene set with its corrected p-value,
  and arrange the table by increasing p-value.

```{r}
go_info <- read_excel("data/GSE111543.xlsx", "go_info")
go_info$BH <- fisher_test_bh[go_info$GOID]
go_info <- arrange(go_info, BH)
head(go_info)
```
